test_ondemand_download::test_ondemand_download_large_rel flakyness

Two ways:
- `assert 7 > 7` on the last line
- pageserver stop failed: pageserver with pid 18039 did not stop in 10 seconds [here](https://neon-github-public-dev.s3.amazonaws.com/reports/main/release/4235314752/index.html#suites/b97efae3a617afb71cb8142f5afa5224/da1f1dd229669df1/)

Noted, didn't fix yet:
- gc and compaction run on the second restart
- comments say that they are disabled, but period is non-zero, so they are executed

From @bayandin's gist:

```
subprocess.CalledProcessError: Command '['/tmp/neon/bin/neon_local', 'pageserver', 'stop']' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:
test_runner/regress/test_ondemand_download.py:110: in test_ondemand_download_large_rel
    env.pageserver.stop()
test_runner/fixtures/neon_fixtures.py:1648: in stop
    self.env.neon_cli.pageserver_stop(immediate)
test_runner/fixtures/neon_fixtures.py:1425: in pageserver_stop
    return self.raw_cli(cmd)
test_runner/fixtures/neon_fixtures.py:1215: in raw_cli
    raise Exception(msg) from subprocess.CalledProcessError(
E   Exception:             Run ['/tmp/neon/bin/neon_local', 'pageserver', 'stop'] failed:
E                 stdout: Stopping pageserver with pid 29195 gracefully.......
E   pageserver has not stopped yet, continuing to wait.....
E
E                 stderr: pageserver stop failed: pageserver with pid 29195 did not stop in 10 seconds
```

- [test_ondemand_download_large_rel[debug-pg14-real_s3]](https://neon-github-public-dev.s3.amazonaws.com/reports/main/4992306427/index.html#suites/b97efae3a617afb71cb8142f5afa5224/e3c53c670ef9d6db/)
- [test_ondemand_download_large_rel[debug-pg14-real_s3]](https://neon-github-public-dev.s3.amazonaws.com/reports/main/4992477497/index.html#suites/b97efae3a617afb71cb8142f5afa5224/91ee43ac1f702854/)
- [test_ondemand_download_large_rel[debug-pg15-real_s3]](https://neon-github-public-dev.s3.amazonaws.com/reports/main/4991720140/index.html#suites/b97efae3a617afb71cb8142f5afa5224/412e68ef15af5ef3/)
- [test_ondemand_download_large_rel[debug-pg15-real_s3]](https://neon-github-public-dev.s3.amazonaws.com/reports/main/4992306427/index.html#suites/b97efae3a617afb71cb8142f5afa5224/8815bba08b8cb5f0/)
- [test_ondemand_download_large_rel[release-pg15-real_s3]](https://neon-github-public-dev.s3.amazonaws.com/reports/main/4992477497/index.html#suites/b97efae3a617afb71cb8142f5afa5224/14be1c02dd3c5553/)
- [test_ondemand_download_large_rel[release-pg14-real_s3]](https://neon-github-public-dev.s3.amazonaws.com/reports/main/4992805591/index.html#suites/b97efae3a617afb71cb8142f5afa5224/2ea9f71b6495032e/)
- [test_ondemand_download_large_rel[release-pg15-real_s3]](https://neon-github-public-dev.s3.amazonaws.com/reports/main/4992805591/index.html#suites/b97efae3a617afb71cb8142f5afa5224/f5ae57dc65391fc9/)

This is a mix of real_s3 trouble #4068 and download_large_rel needing a longer wait time?

--- a/test_runner/regress/test_ondemand_download.py
+++ b/test_runner/regress/test_ondemand_download.py
@@ -101,6 +101,9 @@ def test_ondemand_download_large_rel(
     # wait until pageserver receives that data
     wait_for_last_record_lsn(client, tenant_id, timeline_id, current_lsn)
 
+    # stop pg before checkpoint to stop wal generation
+    endpoint.stop()
+
     # run checkpoint manually to be sure that data landed in remote storage
     client.timeline_checkpoint(tenant_id, timeline_id)
 
@@ -109,7 +112,6 @@ def test_ondemand_download_large_rel(
     log.info("uploads have finished")
 
     ##### Stop the first pageserver instance, erase all its data
-    endpoint.stop()
     env.pageserver.stop()
 
     # remove all the layer files

--- a/test_runner/regress/test_ondemand_download.py
+++ b/test_runner/regress/test_ondemand_download.py
@@ -63,7 +63,7 @@ def test_ondemand_download_large_rel(
     tenant, _ = env.neon_cli.create_tenant(
         conf={
             # disable background GC
-            "gc_period": "10 m",
+            "gc_period": "0s",
             "gc_horizon": f"{10 * 1024 ** 3}",  # 10 GB
             # small checkpoint distance to create more delta layer files
             "checkpoint_distance": f"{10 * 1024 ** 2}",  # 10 MB

--- a/test_runner/regress/test_ondemand_download.py
+++ b/test_runner/regress/test_ondemand_download.py
@@ -67,8 +67,11 @@ def test_ondemand_download_large_rel(
             "gc_horizon": f"{10 * 1024 ** 3}",  # 10 GB
             # small checkpoint distance to create more delta layer files
             "checkpoint_distance": f"{10 * 1024 ** 2}",  # 10 MB
+            # allow compaction with the checkpoint
             "compaction_threshold": "3",
             "compaction_target_size": f"{10 * 1024 ** 2}",  # 10 MB
+            # but don't run compaction in background or on restart
+            "compaction_period": "0s",
         }
     )
     env.initial_tenant = tenant

--- a/test_runner/regress/test_ondemand_download.py
+++ b/test_runner/regress/test_ondemand_download.py
@@ -98,11 +98,16 @@ def test_ondemand_download_large_rel(
 
         current_lsn = Lsn(query_scalar(cur, "SELECT pg_current_wal_flush_lsn()"))
 
-    # wait until pageserver receives that data
     wait_for_last_record_lsn(client, tenant_id, timeline_id, current_lsn)
 
-    # stop pg before checkpoint to stop wal generation
+    # stop endpoint before checkpoint to stop wal generation
     endpoint.stop()
+
+    # stopping of safekeepers now will help us not to calculate logical size
+    # after startup, so page requests should be the only one on-demand
+    # downloading the layers
+    for sk in env.safekeepers:
+        sk.stop()
 
     # run checkpoint manually to be sure that data landed in remote storage
     client.timeline_checkpoint(tenant_id, timeline_id)
@@ -122,8 +127,13 @@ def test_ondemand_download_large_rel(
     ##### Second start, restore the data and ensure it's the same
     env.pageserver.start()
 
-    endpoint.start()
+    # start a readonly endpoint which we'll use to check the database.
+    # readonly (with lsn=) is required so that we don't try to connect to
+    # safekeepers, that have now been shut down.
+    endpoint = env.endpoints.create_start("main", lsn=current_lsn)
+
     before_downloads = get_num_downloaded_layers(client, tenant_id, timeline_id)
+    assert before_downloads != 0, "basebackup should on-demand non-zero layers"
 
     # Probe in the middle of the table. There's a high chance that the beginning
     # and end of the table was stored together in the same layer files with data

